assembly/fabric8-ide-dashboard-war/src/components/ide-fetcher/che-ide-fetcher.service.ts

```diff
@@ -0,0 +1,155 @@
+import {CheBranding} from '../branding/che-branding.factory';
+
+const IDE_FETCHER_CALLBACK_ID = 'cheIdeFetcherCallback';
+
+/**
+ * Provides a way to download IDE .js and then cache it before trying to load the IDE
+ * @author Florent Benoit
+ * @author Oleksii Orel
+ * @author David Festal
+ */
+export class CheIdeFetcher {
+
+  static $inject = ['$log', '$http', '$window', 'cheBranding'];
+
+  private $log: ng.ILogService;
+  private $http: ng.IHttpService;
+  private $window: ng.IWindowService;
+  private cheBranding: CheBranding;
+  private userAgent: string;
+
+  /**
+   * Default constructor that is using resource injection
+   */
+  constructor($log: ng.ILogService, $http: ng.IHttpService, $window: ng.IWindowService, cheBranding: CheBranding) {
+    this.$log = $log;
+    this.$http = $http;
+    this.$window = $window;
+    this.cheBranding = cheBranding;
+
+    this.userAgent = this.getUserAgent();
+
+    const callback = () => {
+      this.findMappingFile();
+      this.cheBranding.unregisterCallback(IDE_FETCHER_CALLBACK_ID);
+    };
+    this.cheBranding.registerCallback(IDE_FETCHER_CALLBACK_ID, callback.bind(this));
+  }
+
+  /**
+   * Gets user agent.
+   * @returns {string}
+   */
+  getUserAgent(): string {
+    const userAgent = this.$window.navigator.userAgent.toLowerCase();
+    const docMode = (<any>this.$window.document).documentMode;
+
+    if (userAgent.indexOf('webkit') !== -1) {
+      return 'safari';
+    } else if (userAgent.indexOf('msie') !== -1) {
+      if (docMode >= 10 && docMode < 11) {
+        return 'ie10';
+      } else if (docMode >= 9 && docMode < 11) {
+        return 'ie9';
+      } else if (docMode >= 8 && docMode < 11) {
+        return 'ie8';
+      }
+    } else if (userAgent.indexOf('gecko') !== -1) {
+      return 'gecko1_8';
+    }
+
+    return 'unknown';
+  }
+
+  /**
+   * Finds mapping file.
+   */
+  findMappingFile(): void {
+    // get the content of the compilation mapping file
+    const randVal = Math.floor((Math.random() * 1000000) + 1);
+    const resourcesPath = this.cheBranding.getIdeResourcesPath();
+    if (!resourcesPath) {
+      this.$log.log('Unable to get IDE resources path');
+      return;
+    }
+
+    const fileMappingUrl = `${resourcesPath}compilation-mappings.txt?uid=${randVal}`;
+
+    this.$http.get(fileMappingUrl).then((response: {data: any}) => {
+      if (!response || !response.data) {
+        return;
+      }
+      let urlToLoad = this.getIdeUrlMappingFile(response.data);
+      // load the url
+      if (angular.isDefined(urlToLoad)) {
+        this.$log.log('Preloading IDE javascript', urlToLoad);
+        this.$http.get(urlToLoad, { cache: true});
+      } else {
+        this.$log.error('Unable to find the IDE javascript file to cache');
+      }
+    }, (error: any) => {
+      this.$log.log('unable to find compilation mapping file', error);
+    });
+
+    this.$http.get('/api/cdn-support/paths', {cache: false}).then((response: {data: any}) => {
```

Comment:
It seems that resources would be loaded even when a non-default version/implementation of the editor is used. Which will make Che slower on start. Is it intentional for our prefetching techniuque?

Reply:
I'm not sure I understand what you mean. We do exactly the same thing as what was done until now with the main GWT cacheable Javascript artifact: pre-fetch it in the Dashboard in all cases.
For now what is considered as the default Che 7 editor is configurable by Che server. In the future it might be configurable by user.

But these resources are the resources of the main Che 7 editor, and the whole idea here is to have them always up-to-date and ready to use in the user browser, even if, at some point in time, the user would decide to use another alternate editor from times to times.

Reply:
In Che 6 we have only one IDE - GWT based one. In Che 7 it is possible to have different IDEs for different users/workspaces or different versions of Theia IDE for different workspaces. Which means that users that use non-default IDE/version of Theia would download resources wastefully. 
I don't say that it is a showstopper for the PR. Overall ppl would use default Theia mostly. Just wanted to know the intentions.

Reply:
Let me come back to your comment here: https://github.com/eclipse/che/issues/11943#issuecomment-439310541, especially this quote:

> And since we are going to decrease the time needed to pull images and start Theia machine we need to load resources earlier than a workspace start.

If I understand correctly, the idea was that the earliest we pre-fetch editor resources, the fastest the editor execution will be at workspace start.
And earliest ==  when User dashboard starts on the browser.

But of course in Che 7 we cannot do that for all editors since the list is not finite.

So let's do it for the default editor for now.

And in the future, let's provide an option to configure this by user: the user would be able to override the default editor setting, possibly with a NULL value, if he doesn't want any default editor resources to be pre-fetched.

Doesn't this make sense ?

Reply:
Idea was to try to load the default editor as fast as possible (before the user even choose a stack)
I think it was discussed that workspace-loader component could do the stuff as well as at that step we already know what is the 'selected' editor but during the time the user select a stack/workspace we could already have cached the default editor locally.


Also one major difference is that instead of downloading each time theia JS files (as the endpoint is changing) now it's always cached.
And each time a user come back, if editor has not changed, JS will be the same so there is no big overhead. Also it's done in rh-che project (which promote kind of Theia as default editor), not che.